src/util/detect.changes.ts

```diff
@@ -0,0 +1,127 @@
+import { isEqual } from 'lodash';
+import { ExcalidrawElement } from '../excalidraw/types';
+
+type Identifiable = {
+  id: string;
+};
+
+type Changes<T extends Identifiable> = {
+  id: string;
+} & Partial<{
+  [P in keyof T as P extends 'id' ? never : P]: {
+    old?: T[P];
+    new?: T[P];
+  };
+}>;
+
+export type DetectedChanges<T extends Identifiable> = {
+  inserted?: Array<T>;
+  updated?: Array<Changes<T>>;
+  deleted?: Array<{
+    id: string;
+  }>;
+};
+
+export type DetectedChangesType = 'insert' | 'update' | 'delete' | 'unknown';
+
+// todo: doc
+/**
+ * Detects deep changes between two arrays of ExcalidrawElements, and returns an object with the changes.
+ * Returns null if no changes are detected.
+ * @param oldArrInput
+ * @param newArrInput
+ * @param fieldsToIgnore
+ */
+export const detectChanges = (
```

Comment:
I believe this method is the essence of the PR, it is algorithmic, and it merits writing unit tests (even if most of those are auto-generated with GenAI).

Reply:
added